bpfd/src/bpf.rs

```diff
@@ -0,0 +1,358 @@
+                return Err(BpfdError::InvalidAttach(program.info.to_string()));
+            }
+            let category = parts[0].to_owned();
+            let name = parts[1].to_owned();
+
+            let map_pin_path = format!("{RTDIR_FS_MAPS}/{id}");
+            fs::create_dir_all(map_pin_path.clone())
+                .map_err(|e| BpfdError::Error(format!("can't create map dir: {}", e)))?;
+
+            let mut loader = BpfLoader::new()
+                .map_pin_path(map_pin_path.clone())
+                .load_file(&program.data.path)?;
+
+            let tracepoint: &mut TracePoint = loader
+                .program_mut(&program.data.section_name)
+                .ok_or_else(|| {
+                    let _ = fs::remove_dir_all(map_pin_path);
+                    BpfdError::SectionNameNotValid(program.data.section_name.clone())
+                })?
+                .try_into()?;
+
+            tracepoint.load()?;
+            p.save(id)
+                .map_err(|_| BpfdError::Error("unable to persist program data".to_string()))?;
+            self.programs.insert(id, p);
+
+            tracepoint.attach(&category, &name)?;
+            tracepoint
+                .pin(format!("{RTDIR_FS}/prog_{id}_link"))
+                .map_err(|_| BpfdError::UnableToPin)?;
+            Ok(id)
+        } else {
+            panic!("not a tracepoint program")
+        }
+    }
+
+    pub(crate) fn remove_program(&mut self, id: Uuid, owner: String) -> Result<(), BpfdError> {
+        if let Some(prog) = self.programs.get(&id) {
+            if !(prog.owner() == &owner || owner == SUPERUSER) {
+                return Err(BpfdError::NotAuthorized);
+            }
+        } else {
+            return Err(BpfdError::InvalidID);
+        }
+
+        let prog = self.programs.remove(&id).unwrap();
+
+        prog.delete(id)
+            .map_err(|_| BpfdError::Error("unable to delete program data".to_string()))?;
+
+        match prog {
+            Program::Xdp(_) | Program::Tc(_) => self.remove_multi_attach_program(prog),
+            Program::Tracepoint(_) => Ok(()),
+        }
+    }
+
+    pub(crate) fn remove_multi_attach_program(
+        &mut self,
+        program: Program,
+    ) -> Result<(), BpfdError> {
+        // Calculate the next_available_id
+        let next_available_id = self
+            .programs
+            .iter()
+            .filter(|(_, p)| {
+                if p.kind() == program.kind() {
+                    p.if_index() == program.if_index() && p.direction() == program.direction()
+                } else {
+                    false
+                }
+            })
+            .collect::<HashMap<_, _>>()
+            .len();
+
+        let did = program
+            .dispatcher_id()
+            .ok_or(BpfdError::DispatcherNotRequired)?;
+
+        let old_dispatcher = self.dispatchers.remove(&did);
+
+        if let Some(ref old) = old_dispatcher {
+            if next_available_id == 0 {
```

Comment:
When the last tc program is deleted, the tc dispatcher doesn't get detached.  So, when we hit this condition, for a tc dispatcher, we need to explicitly call detach().  However, I haven't been able to figure out how to access the link so I can call link.detach().

Reply:
I've fixed this now - fix fixed in `TcDispatcher::delete` where we
```rust
if let Some(link) = self.link.take() {
    link.detach()?;
}
```